src/Confluent.Kafka/Diagnostics.cs

```diff
@@ -0,0 +1,118 @@
+﻿// Copyright 2016-2018 Confluent Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Refer to LICENSE for more information.
+
+using System;
+using System.Diagnostics;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Text;
+
+namespace Confluent.Kafka
+{
+    internal static class Diagnostics
+    {
+        public const string DiagnosticListenerName = "Confluent.Kafka.Listener";
+        public const string TraceParentHeaderName = "traceparent";
+        public const string TraceStateHeaderName = "tracestate";
+
+        public static DiagnosticListener DiagnosticSource { get; } = new DiagnosticListener(DiagnosticListenerName);
+
+        public static class Producer
+        {
+            public const string ActivityName = DiagnosticListenerName + ".MessageProduced";
+            public const string StartName = ActivityName + ".Start";
+            public const string StopName = ActivityName + ".Stop";
+            public const string ExceptionName = ActivityName + ".Exception";
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static Activity Start<TKey, TValue>(string topic, Message<TKey, TValue> message)
+            {
+                if (DiagnosticSource.IsEnabled(ActivityName, topic))
+                {
+                    Activity activity = new Activity(ActivityName);
+                    activity.Start();
+
+                    AddActivityStateToHeaders(activity, message.Headers);
+
+                    if (DiagnosticSource.IsEnabled(StartName))
+                    {
+                        DiagnosticSource.Write(StartName, new { Topic = topic, Message = message });
```

Comment:
Instead of writing the new anonymous type, (thus requiring adapters to have knowledge about this type etc), the new recommendation would be to use `activity.AddTag("messaging.system", "kafka")`

OpenTelemetry has conventions for the tag names, which can be used:
https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/messaging.md